### ISSUE TITLE
Less-than-helpful "Undefined index: error" message.

### ISSUE THREAD

**Issue by cviebrock** [elastic/elasticsearch-php#245]
Take the following code:

```php
$client = \Elasticsearch\ClientBuilder::create()->build();
$params = [
	'index' => 'foo',
	'body' => [
		'settings' => [],
		'mappings' => [
			'bar' => [
				'properties' => [
					'first_name' => [
						'type' => 'string',
					],
					'age' => [
						'type' => 'integer'
					]
				]
			]
		]
	]
];

$ret = $client->indices()->create($params);
```

This will fail with "Undefined index: error" in `vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php` on line 552.  The reason being that the $response array that comes back has a _body_ of:

```json
{"error":"ClassCastException[java.util.ArrayList cannot be cast to java.util.Map]","status":500}
```

but no _error_ index.

The problem is my empty "settings" array, but I'm thinking that the client shouldn't fail this hard without some more descriptive message?  Maybe line 552 of that class just needs to be wrapped in a `array_key_exists()` check? Maybe 500 errors from the server don't set that _error_ index in the reponse?

Thanks!

**Comment by polyfractal** [comment]
Oh yeah, this looks like a legit bug.  :)

The "error" index is supposed to be populated with the curl error string (not an error the body itself, although it confusingly looks like it should). It appears I forgot to port that piece of code during the 2.0 rewrite.

Thanks for the report, I'll try to get this fixed soon (likely next week some time, I'm pretty swamped this week).

**Comment by cviebrock** [comment]
No problem.  Happy to help! :)

**Comment by stha** [comment]
Is there already a plan when to fix this? 

I'm getting a `Notice: Undefined index: error in vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php on line 552`, which is a little bit disturbing.

**Comment by polyfractal** [comment]
Thanks for the nudge, this fell off my radar due to other projects.  Just pushed a fix: f22c6c209e630376081f0d66e13f2c9040feab53  I want to batch together a few more changes in ES 2.0 then I'll tag it, probably in a few hours.

I also took the opportunity to add support for 2.0 structured exceptions.  Post ES 2.0, you'll see messages like this:

```
Elasticsearch\Common\Exceptions\ServerErrorResponseException : class_cast_exception: java.util.ArrayList cannot be cast to java.util.Map
```

And if you interrogate the exception itself, the root exception will hold a nice JSON error that you can decode yourself.

Pre-2.0, we are still stuck with semi-structured (and sometimes completely unstructured) exceptions.  The resulting error will look more like this:

```
Elasticsearch\Common\Exceptions\ServerErrorResponseException : ClassCastException[java.util.ArrayList cannot be cast to java.util.Map]
```

But will be a lot more variable, since each exception stack coming back from ES tends to be a unique snowflake.  And sometimes just a big stack dump with nothing resembling a human readable exception.

**Comment by polyfractal** [comment]
Just tagged `v2.0.0-beta5` which includes this fix :)
